### PR TITLE
Update XGBoost to 1.0

### PR DESCRIPTION
**Pull request by balajitummala**
*Issue #, if available:*

*Description of changes:*

XGBoost update to 1.0

- Modified the Dockerfile to update xgboost to 1.0
- Added validation support and test cases for new parameters introduced in 1.0:
  montone_constraints and interaction_constraints
- Updated 'tree updater' options

Release Notes: https://xgboost.readthedocs.io/en/release_1.0.0/

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by iyerr3** [comment]
Top level comment: Please add summary of changes into 1.0 including links to upstream release notes and backwards-incompatible changes.

File: src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py [hunk 1]
```diff
@@ -98,6 +99,19 @@ def eval_metric_dep_validator(value, dependencies):
                     'binary:', 'rank:']):
                 raise exc.UserError("Metric 'auc' can only be applied for classification and ranking problems.")
 
+    @hpv.dependencies_validator(["tree_method"])
+    def monotone_constraints_validator(value, dependencies):
+        if value is not None and dependencies.get("tree_method") not in ("exact", "hist"):
```
Comment (aws-patlin) [inline src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py:104]: For readability, can you follow the same pattern in `interaction_constraints_validator`? i.e.
```
tree_method = dependencies.get("tree_method")
if value is not None and tree_method not in ("exact", "hist"):
```

File: src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py [hunk 2]
```diff
@@ -98,6 +99,19 @@ def eval_metric_dep_validator(value, dependencies):
...
+            raise exc.UserError("Montonic constraints can be used only when the tree method parameter is set to "
```
[the remark this answers is not in the document]
Comment (aws-patlin) [reply src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py:105]: Also, let's use the actual parameter names: "'monotone_constraints' can be used only when the 'tree_method' parameter is set to either 'exact' or 'hist'."

File: src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py [hunk 3]
```diff
@@ -98,6 +99,19 @@ def eval_metric_dep_validator(value, dependencies):
...
+                                "either 'exact' or 'hist'.")
+
+    @hpv.dependencies_validator(["tree_method"])
+    def interaction_constraints_validator(value, dependencies):
+        tree_method = dependencies.get("tree_method")
+        if value is not None and tree_method not in ("exact", "hist", "approx"):
+            raise exc.UserError("Interaction constraints can be used only when the tree method parameter is set to "
```
Comment (aws-patlin) [inline src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py:112]: Same as above, let's use the actual parameter names.

File: src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py [hunk 4]
```diff
@@ -188,7 +206,7 @@ def eval_metric_dep_validator(value, dependencies):
                                       range=["binary:logistic", "binary:logitraw", "binary:hinge",
                                              "count:poisson", "multi:softmax", "multi:softprob",
                                              "rank:pairwise", "rank:ndcg", "rank:map", "reg:linear",
```
Comment (aws-patlin) [inline src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py:208]: `reg:linear` was removed in 1.0 (`reg:squarederror` is equivalent)

File: tox.ini [hunk 5]
```diff
@@ -20,7 +21,7 @@ conda_channels=
     conda-forge
     mlio
 commands =
-    pytest --cov=sagemaker_xgboost_container --cov-fail-under=60 test/unit # increase minimum bar over time (75%+)
+    pytest --cov=sagemaker_xgboost_container --cov-fail-under=60 {posargs} # increase minimum bar over time (75%+)
```
Comment (aws-patlin) [inline tox.ini:24]: Does just running `tox` still run unit tests, or do we have to specify `tox test/unit` after this change?

Comment (balajitummala) [reply tox.ini:24]: We need to specify test/unit, added this to the README as well.

File: src/sagemaker_algorithm_toolkit/hyperparameter_validation.py [hunk 6]
```diff
@@ -176,6 +177,50 @@ def validate_range(self, value):
...
+            return eval(value)
+        elif isinstance(value, list) or isinstance(value, tuple):
+            return value
+
+    def format_range(self):
+        return {"TupleParameterRangeSpecification": {
+            "Values": self.range}}
+
+    def validate_range(self, value):
+        if any([index not in self.range for index in value]):
```
Comment (ericangelokim) [inline src/sagemaker_algorithm_toolkit/hyperparameter_validation.py:220]: Nit: ```index``` is a misleading variable name here, prefer ```element``` or something more clear

File: src/sagemaker_algorithm_toolkit/hyperparameter_validation.py [hunk 7]
```diff
@@ -176,6 +177,50 @@ def validate_range(self, value):
...
+class TupleHyperparameter(Hyperparameter):
+    def __init__(self, *args, **kwargs):
+        if kwargs.get("range") is None:
+            raise exc.AlgorithmError("range must be specified")
+        super(TupleHyperparameter, self).__init__(*args, **kwargs)
+
+    def parse(self, value):
+        if isinstance(value, str):
+            return eval(value)
+        elif isinstance(value, list) or isinstance(value, tuple):
```
Comment (ericangelokim) [inline src/sagemaker_algorithm_toolkit/hyperparameter_validation.py:212]: I'm not convinced on the correctness of the design here. Shouldn't we fail validation if the value is a list?

Comment (balajitummala) [reply src/sagemaker_algorithm_toolkit/hyperparameter_validation.py]: The [documentation](https://xgboost.readthedocs.io/en/latest/tutorials/monotonic.html) about montone constraints is not clear regarding the accepted format.
This was [where](https://github.com/dmlc/xgboost/issues/1514) I found out that python supports tuple format but this issue was raised in 2016 so added 'list' to not fail a validation.
I modified it to support only 'tuple' in the new commit.

File: README.rst [hunk 8]
```diff
@@ -59,7 +59,7 @@ Base Images
 The "base" Dockerfile encompass the installation of the framework and all of the dependencies
 needed.
 
-Tagging scheme is based on <SageMaker-XGBoost-version>-cpu-py3 (e.g. 0.90-2-cpu-py3), where
+Tagging scheme is based on <SageMaker-XGBoost-version>-cpu-py3 (e.g. 1.0-1-cpu-py3), where
```
[the remark this answers is not in the document]
Comment (balajitummala) [reply README.rst]: Added variable substitution to README

File: README.rst [hunk 9]
```diff
@@ -31,6 +31,8 @@ development machine:
 
 - `Docker <https://www.docker.com/>`__
 
+Note: CMake is required for XGBoost. If using macOS, install CMake (pip install cmake)
```
Comment (aws-patlin) [inline README.rst:34]: Would it be better for Mac users to install CMake via brew? `brew install cmake`

Comment (balajitummala) [reply README.rst:34]: pip comes default with macOS so chose pip over brew

File: src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py [hunk 10]
```diff
@@ -187,8 +206,8 @@ def eval_metric_dep_validator(value, dependencies):
         hpv.CategoricalHyperparameter(name="objective",
                                       range=["binary:logistic", "binary:logitraw", "binary:hinge",
                                              "count:poisson", "multi:softmax", "multi:softprob",
-                                             "rank:pairwise", "rank:ndcg", "rank:map", "reg:linear",
-                                             "reg:squarederror", "reg:logistic", "reg:gamma",
+                                             "rank:pairwise", "rank:ndcg", "rank:map", "reg:linear", "reg:squarederror",
```
Comment (iyerr3) [inline src/sagemaker_xgboost_container/algorithm_mode/hyperparameter_validation.py:209]: Retain `"reg:squarederror"` in the new line

File: README.rst [hunk 11]
```diff
@@ -159,8 +159,12 @@ If you want to run unit tests, then use:
     # or you can use tox to run unit tests as well as flake8 and code coverage
 
     tox
-    tox -e py3-xgboost0.90,flake8
-    tox -e py3-xgboost0.72,py3-xgboostlatest
+    tox -e py3-xgboost1.0,flake8 test/unit
+    tox -e py3-xgboost0.90,py3-xgboostlatest
```
[the remark this answers is not in the document]
Comment (balajitummala) [reply README.rst:165]: Added back previous version.

**Pull request merged**